google/cloud/security/common/gcp_api/_base_repository.py

```diff
@@ -0,0 +1,713 @@
+        OAuth2Credentials: Credentials as built by oauth2client.
+
+    Raises:
+        api_errors.ApiExecutionError: If fails to build credentials.
+    """
+    try:
+        credentials = (
+            service_account.ServiceAccountCredentials.from_json_keyfile_name(
+                keyfile_name, scopes=scopes))
+    except (ValueError, KeyError, TypeError, IOError) as e:
+        raise api_errors.ApiExecutionError(
+            'Error building admin api credential: %s', e)
+
+    return credentials.create_delegated(delegated_account)
+
+
+def flatten_list_results(paged_results, item_key):
+    """Flatten a split-up list as returned by list_next() API.
+
+    GCE 'list' APIs return results in the form:
+      {item_key: [...]}
+    with one dictionary for each "page" of results. This method flattens
+    that to a simple list of items.
+
+    Args:
+        paged_results (list): A list of paged API response objects.
+            [{page 1 results}, {page 2 results}, {page 3 results}, ...]
+        item_key (str): The name of the key within the inner "items" lists
+            containing the objects of interest.
+
+    Returns:
+        list: A list of items.
+    """
+    results = []
+    for page in paged_results:
+        results.extend(page.get(item_key, []))
+    return results
+
+
+def flatten_aggregated_list_results(paged_results, item_key):
+    """Flatten a split-up list as returned by GCE "aggregatedList" API.
+
+    The compute API's aggregatedList methods return a structure in
+    the form:
+      {
+        items: {
+          $group_value_1: {
+            $item_key: [$items]
+          },
+          $group_value_2: {
+            $item_key: [$items]
+          },
+          $group_value_3: {
+            "warning": {
+              message: "There are no results for ..."
+            }
+          },
+          ...,
+          $group_value_n, {
+            $item_key: [$items]
+          },
+        }
+      }
+    where each "$group_value_n" is a particular element in the
+    aggregation, e.g. a particular zone or group or whatever, and
+    "$item_key" is some type-specific resource name, e.g.
+    "backendServices" for an aggregated list of backend services.
+
+    This method takes such a structure and yields a simple list of
+    all $items across all of the groups.
+
+    Args:
+        paged_results (list): A list of paged API response objects.
+            [{page 1 results}, {page 2 results}, {page 3 results}, ...]
+        item_key (str): The name of the key within the inner "items" lists
+            containing the objects of interest.
+
+    Returns:
+        list: A list of items.
+    """
+    items = []
+    for page in paged_results:
+        aggregated_items = page.get('items', {})
+        for items_for_grouping in aggregated_items.values():
+            for item in items_for_grouping.get(item_key, []):
+                items.append(item)
+    return items
+
+
+class BaseRepositoryClient(object):
```

Comment:
General comment:
This class feels like a factory class for GCP Repositories. Furthermore, it is not intuitive to have abstractions for both: Repositories and RepositoryClients. I would suggest (in future releases) to consider transitioning BaseRepositoryClient into a factory class for GCP Repositories.